### PR TITLE
Fix structural eq of lambdas with diff loc

### PR DESCRIPTION
**Pull request by martong**
This PR is dependent on #646 and #647. So first please review them, once the dependencies merged I'll rebase this PR.

### CODE REVIEW

File: unittests/AST/ASTImporterTest.cpp [hunk 1]
```diff
@@ -4677,6 +4677,79 @@ TEST_P(ASTImporterOptionSpecificTestBase, LambdaInFunctionParam) {
...
+  CXXRecordDecl *FromL0 =
+      FirstDeclMatcher<CXXRecordDecl>().match(FromTU, Pattern);
+  CXXRecordDecl *FromL1 =
+      LastDeclMatcher<CXXRecordDecl>().match(FromTU, Pattern);
+  ASSERT_NE(FromL0, FromL1);
+
+  Import(FromL0, Lang_CXX11);
+  Import(FromL1, Lang_CXX11);
+  Decl *ToTU = ToAST->getASTContext().getTranslationUnitDecl();
+  CXXRecordDecl *ToL0 = FirstDeclMatcher<CXXRecordDecl>().match(ToTU, Pattern);
```
Comment (balazske) [inline unittests/AST/ASTImporterTest.cpp]: We can get the `ToL0` and `ToL1` from the result of the `Import` calls.

File: unittests/AST/StructuralEquivalenceTest.cpp [hunk 2]
```diff
@@ -838,6 +838,33 @@ TEST_F(StructuralEquivalenceLambdaTest, LambdaClassesWithEqFields) {
...
+      R"(
+      #define LAMBDA [](){}
+      void f() { auto L =       LAMBDA; }
+      )",
+      Lang_CXX11, lambdaExpr(), lambdaExpr());
+  CXXRecordDecl *L0 = get<0>(t)->getLambdaClass();
+  CXXRecordDecl *L1 = get<1>(t)->getLambdaClass();
+  EXPECT_FALSE(testStructuralMatch(L0, L1));
+}
+
```
Comment (balazske) [inline unittests/AST/StructuralEquivalenceTest.cpp:867]: These last 2 tests look a bit strange, these could be equivalent too. Why is this needed? Probably it is OK to make lambdas never structurally equivalent?

Comment (martong) [reply unittests/AST/StructuralEquivalenceTest.cpp:867]: According to the standard, every new lambda expression introduces a new type.
So inside one TU, every different lambda should have a different type, yes. However, across TUs, if the lambda is coming from a header with the very same location then I think we should not create a new type, because that is essentially the very same lambda. We can satisfy both if we check for the location, can't we?

File: unittests/AST/ASTImporterTest.cpp (hunk 1 above)
Comment (martong) [reply unittests/AST/ASTImporterTest.cpp]: That is a good point, thanks.

File: unittests/AST/StructuralEquivalenceTest.cpp (hunk 2 above)
Comment (balazske) [reply unittests/AST/StructuralEquivalenceTest.cpp:867]: Normally it should work. If an imported TU uses a copy of a header (with exact same contents or re-formatted) that is used by the already imported things it may not work. And theoretically two code parts where only the amount of whitespace is different should be equivalent.
Probably it is possible to use a similar check like at unnamed structs: Check the position of the lambda Decl in its parent (not sure if it is good always)?

**Review by balazske — APPROVED** [review]

**Comment by martong** [comment]
https://reviews.llvm.org/D64078

**Comment by martong** [comment]
Alternative patch https://reviews.llvm.org/D66348
